nncf/initialization.py

```diff
@@ -164,39 +164,66 @@ def __init__(self, model, init_device: str, num_bn_forget_steps):
         self.num_bn_forget_steps = num_bn_forget_steps
         self.momentum_bn_forget = 0.9
         self.original_momenta_values = {}
+        self.original_training_state = {}
 
     @staticmethod
     def _apply_to_batchnorms(func):
         def func_apply_to_bns(module):
-            if isinstance(module, torch.nn.modules.batchnorm.BatchNorm2d):
+            if isinstance(module, (torch.nn.modules.batchnorm.BatchNorm1d,
+                                   torch.nn.modules.batchnorm.BatchNorm2d,
+                                   torch.nn.modules.batchnorm.BatchNorm3d)):
                 func(module)
 
         return func_apply_to_bns
 
-    def _run_model_inference(self, data_loader, num_init_steps, device):
-        num_bn_forget_steps = self.num_bn_forget_steps
+    @contextmanager
+    def _bn_training_state_switcher(self):
+        def save_original_bn_training_state(module):
+            self.original_training_state[module] = module.training
+
+        def set_bn_training_state(module, state):
+            module.training = state
+
+        def restore_original_bn_training_state(module):
+            module.training = self.original_training_state[module]
+
```

Comment:
Why isn't using the modified `training_mode_switcher` enough?

Reply:
The approach using modified `training_mode_switcher` is not fully correct. In BatchNorm adaptation algorithm we only want to update BatchNorm statistics, while switching the whole model in training mode can also affect on other model's parts.

Reply:
I think `training_mode_switcher` could be sufficient in practice, but switching BN `training` flags only is a safer approach, since we potentially preserve inference-time behavior of other layers such as DropOut while refreshing BN statistics.